### PR TITLE
Initial commit of iam_managed_policy module

### PR DESCRIPTION
**Pull request by dkhenry**
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

iam_managed_policy

##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY

This module can add, remove, update versions, and set default versions
of managed policies. It will cycle out old versions of policies if too
many are present. It will check and set the version of the policy that
matches the passed in policy document if one already exists.

### CODE REVIEW

**Comment by mattclay** [comment]
CI failure due to a PEP 8 issue:

```
2017-02-28 16:46:32 ERROR: PEP 8: lib/ansible/modules/cloud/amazon/iam_managed_policy.py:245:14: E711 comparison to None should be 'if cond is None:' (current)
```

The PEP 8 tests can be run locally with `make pep8`.

File: lib/ansible/modules/cloud/amazon/iam_managed_policy.py
Comment (gundalow) [inline lib/ansible/modules/cloud/amazon/iam_managed_policy.py]: Please review docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html
* `descriptions:` should be full sentences, inc full stops
* `required: false` is implicit and not needed
* The examples don't need `task:` they should be formatted so someone can copy and paste them into an existing playbook

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/iam_managed_policy.py
Comment (dkhenry) [reply lib/ansible/modules/cloud/amazon/iam_managed_policy.py]: descriptions and required flags should be addressed. Also I have removed the task flags, and added in an author and requirements block

**Review by gundalow — APPROVED** [review]

**Comment by harsha-y** [comment]
Hello any chance that this PR will be reviewed/merged into ansible_2.3? Anything else blocking this apart from waiting on previous maintainers?

Managed policies is something that hasn't seen support since [Dec 20th, 2015](https://github.com/ansible/ansible-modules-core/issues/2688)

bot_broken

**Comment by gundalow** [comment]
Hi,
This is a new module so once it gets thumbs up it will be merged into Ansible 2.4. We don't backport new features after the `stable-2.3` branch is created.

**Comment by dkhenry** [comment]
Updated this module to pass the 2.4 CI

File: lib/ansible/modules/cloud/amazon/iam_managed_policy.py
Comment (wimnat) [inline lib/ansible/modules/cloud/amazon/iam_managed_policy.py]: Use the helper function in module utils for comparison. See AWS guidelines for details.

**Review by wimnat — CHANGES_REQUESTED** [review]

**Comment by dkhenry** [comment]
@wimnat I have changed it over to use the helper function. Also I have rebased and squashed to a singe commit

**Review by wimnat — COMMENTED** [review]
My other comment would be is it necessary to have both a policy_document and policy_json?

If you set the type to 'json' in your argument spec then you should be able to get away with one parameter. 

E.g.

```
policy: "{{ lookup('file', 'mypolicy') }}"
```

Or

```
policy: "Json string"
```

**Comment by dkhenry** [comment]
@wimnat I wasn't able to get the json type to work when I was developing the module. I didn't know if the types were new in 2.3 or if I was just doing it wrong. The only way I could get it to work was by passing in two parameters. I am willing to try again now that 2.3 is released and I can upgrade my environment.

**Comment by wimnat** [comment]
@dkhenry yeh I think that'd be good. Ping me if you still have issues

**Comment by dkhenry** [comment]
This module has been updated to incorporate feedback. The interface is now to pass in json directly via a policy option 

@wimnat

**Comment by willthames** [comment]
Closing and reopening for a hopefully final CI check

**Comment by willthames** [comment]
Merged, thanks @dkhenry for this!

**Comment by ColOfAbRiX** [comment]
Hi @dkhenry, 

I'm having an issue testing this module. I keep getting "There is an issue in the code of the module. You must specify either both, resource or client to the conn_type parameter in the boto3_conn function call". Another guy had a similar issue and he solved it updating boto, but for me this solution didn't work.

```
$ sudo pip show boto3
Name: boto3
Version: 1.4.4
```

**Comment by dkhenry** [comment]
@ColOfAbRiX can you create an issue, put the full output of the error on it, and tag me

**Comment by ColOfAbRiX** [comment]
I found the issue, it was with my configuration. In a test I ended up having
```
aws_access_key: !!null
...
- iam_managed_policy:
     aws_access_key: "{{ aws_access_key | default(omit) }}"
```
Commenting out the variable ```aws_access_key``` solved the issue.
Thank you anyway

**Pull request merged**
